### DISCUSSION TITLE
Is there a way to prevent large deformations in BSpline transform?

### DISCUSSION THREAD

**Discussion by dzenanz** [SuperElastix/elastix#971] [General]
BrainsFit has [maxBSplineDisplacement](https://slicer.readthedocs.io/en/latest/user_guide/modules/brainsfit.html#advanced-optimization-settings) parameter. Elastix' most similar parameter by name is `DisplacementMagnitudePenalty`, which does something completely different.

I already use `TransformBendingEnergyPenalty`, and it is helpful, but with very finely spaced control grid, I sometimes run into gradual growth/shrinking of some structures. In the screenshots below, left window is reference segmentation, right window is segmentation transferred by registration from another case. This happens only sometimes. What is the best way to deal with it?

<img width="1748" alt="Screenshot 2023-10-11 15 24 02" src="https://github.com/SuperElastix/elastix/assets/1792121/edbe963e-fe0a-4356-b3c1-a75e2c6c8c01">
Vertebrae are compressed in SI direction, and enlarged in other two axes.

<img width="1751" alt="Screenshot 2023-10-11 15 25 36" src="https://github.com/SuperElastix/elastix/assets/1792121/4fe1d1a9-550e-4c90-acce-1dc854b61021">
Face is squished in the LR axis.

**Comment by dzenanz** [comment]
How to add "incompressibility" and "rigidity" penalties described in [elastix 2010 TMI paper](https://elastix.lumc.nl/marius/downloads/2010_j_TMI.pdf)?

**Comment by dzenanz** [comment]
Here is the other failing pair. The problematic deformations are rather localized.
<img width="1005" alt="Screenshot 2023-10-13 11 17 18" src="https://github.com/SuperElastix/elastix/assets/1792121/c9478cef-f314-4c63-aad5-2628257b5fc5">

**Comment by N-Dekker** [comment]
Indeed, elastix does not have such a feature, like "maxBSplineDisplacement". Sorry 🤷 Marius (@mstaring) suggested doing a multi-resolution registration, ensure that the grid size is not too fine, and use a small step size during optimization. Does that help you?

**Reply by dzenanz** [reply]
I will try to limit step size. I am already doing multi-resolution. Would it make sense to increase number of pyramid levels for the stage with finer grid?

**Reply by dzenanz** [reply]
Unfortunately, limiting maximum step size did not help in this scenario.
